PLEASE DON'T MERGE - Just need this on staging (#629)
100mslive/100ms-docs: docs/javascript/v2/guides/javascript-quickstart.mdx

nav: 2.1
---

## Overview

Hello there! In this guide, we'll walk you through step-by-step instructions to create a  video conferencing application using the 100ms JavaScript SDK and test it locally on your device. In case you plan to use React, do check our [React quickstart](https://www.100ms.live/docs/javascript/v2/guides/react-quickstart) as well.

> TL;DR - You can find the link to the complete example over [here](https://codesandbox.io/s/hms-js-quickstart-5rmes).

Reviewer: using "here" or "click here" to create hyperlinks hides an opportunity to provide the user with more useful info. e.g. this can be reframed as 

```suggestion
TL;DR Try out this complete example in [CodeSandbox](https://codesandbox.io/s/hms-js-quickstart-5rmes).
```

## Setup a local Javascript app
In this walkthrough, we’ll use ParcelJS a web application bundler that requires zero configuration. If you wish to use any other library or framework, you can set up your app using that. 

> ### Initialize a project

Reviewer: this is missing steps for installing the 100ms sdk

You can initialize a project using `npm init` or `yarn init`. While initializing the project you can enter the details for your project as you wish.

### Install Parcel
Then, install Parcel into your app using Yarn `yarn add --dev parcel` or when using npm run `npm install --save-dev parcel`.

### Project setup

> Now that parcel is installed, let’s create basic source files for our video-conferencing application. Let’s create the files and folders in the below manner:

Reviewer: I noticed that we're using "parcel" in lower-case here. this is a common problem in our docs - there are lots of places where we use "react" instead of "React", "webrtc" instead of WebRTC. can these be caught by the linting work that we're doing?

```
- index.html
- src
	-index.js
	-styles.css
```

## Add basic functionalities

### Initialize the SDK

Let us start with initializing the libraries. We need an instance of `HMSStore` and `HMSActions` to get started. Copy the code snippet below and paste it into your project.

```jsx:script.js

> import {

Reviewer: the filename needs to be updated here

Reviewer: to index.js

Reviewer: also for other snippets in the page

  HMSReactiveStore,
  selectIsLocalAudioEnabled,
  selectIsLocalVideoEnabled,
  selectPeers,
  selectIsConnectedToRoom
} from "@100mslive/hms-video-store";

// Initialize HMS Store
const hmsManager = new HMSReactiveStore();
hmsManager.triggerOnSubscribe();
const hmsStore = hmsManager.getStore();
const hmsActions = hmsManager.getHMSActions();
```

> `hmsStore` will hold the complete state of the application such as details of all the participants. We can also visualize this state at any time using the

Reviewer: can we link to the docs where HMSStore and Actions are detailed out? e.g. https://100ms-docs-git-dontmergejusttesting-100mslive.vercel.app/docs/api-reference/javascript/v2/interfaces/HMSActions

> -   `Peer`: A participant in the video call. You are the `local peer` while others are `remote peers`.

Reviewer: can we come up with a better definition for Room? the current description doesn't give me much

-   `Track`: Media. There are two types of track a peer can have - audio and video.

### Initialize the HTML elements
Now, let's initialize all the HTML elements required such as tile for local and remote peers, join and leave button, mute and unmute button, etc.

```jsx:script.js
// HTML elements
const form = document.getElementById("join");
const joinBtn = document.getElementById("join-btn");
const conference = document.getElementById("conference");
const peersContainer = document.getElementById("peers-container");
const leaveBtn = document.getElementById("leave-btn");
const muteAud = document.getElementById("mute-aud");
const muteVid = document.getElementById("mute-vid");
const controls = document.getElementById("controls");
```

### Join/leave a room

> To join a room (a video call), we need to call the `join` method on `hmsActions` and it requires us to pass a `config` object.

Reviewer: can we point to the docs for this method in the API reference?

Reviewer: we can use this quickstart as an opportunity to train the user to use the reference section

<Tab id="join-form-1">

```jsx:script.js
// Joining the room
joinBtn.onclick = () => {
  hmsActions.join({
    userName: document.getElementById("name").value,
    authToken: document.getElementById("token").value
  });
};

// Leaving the room
function leaveRoom() {
  hmsActions.leave();
}

// Cleanup if user refreshes the tab or navigates away
window.onunload = window.onbeforeunload = leaveRoom;
leaveBtn.onclick = leaveRoom;
```
</Tab>

![Join room](/guides/join-room.png)

Congratulations! **you now have an audio-only conference ready to use**. If you have a friend join from another device, you can have a nice friendly chat with them.

> ### Render Local and remote peers

Reviewer: inconsistent casing

Reviewer: also feels like this should be called "Render video tiles for peers"

<Tab id="mute-unmute-0">

```jsx:script.js

> //Mute and unmute audio

Reviewer: spacing after `//`

We fetch the current state of the local audio and video and toggle them whenever the buttons are clicked.

![Mute/Unmute](/guides/mute-unmute.gif)

### Changing UI based on connection state

Right now, our join form shows even after we have joined the room. We need a way to know the connection state of the room and hide the form after we've connected.

We can do this by subscribing to the store with the `selectIsConnectedToRoom` selector.

```jsx:script.js
//Showing the required elements on connection/disconnection
function onConnection(isConnected) {
  if (isConnected) {
    form.classList.add("hide");
    conference.classList.remove("hide");
    leaveBtn.classList.remove("hide");
    controls.classList.remove("hide");
  } else {
    form.classList.remove("hide");
    conference.classList.add("hide");
    leaveBtn.classList.add("hide");
    controls.classList.add("hide");
  }
}

// listen to the connection state
hmsStore.subscribe(onConnection, selectIsConnectedToRoom);
```

> ## Test the app

Reviewer: this section seems incomplete?

Reviewer: maybe this is where ParcelJS will be used? not seeing any mention of it in this page besides installation

> ## Next steps

Reviewer: only the last line of this section feels apt for the title "next steps"

Here's the complete example.

<Codesandbox id="5rmes" />

> 

Reviewer: seeing a broken image on the top left of the sandbox 

<img width="327" alt="Screenshot 2022-07-02 at 4 17 59 PM" src="https://user-images.githubusercontent.com/284612/177018733-2bb5ffee-f5ad-44ea-9d42-cddd40579955.png">
